src/librustc_trans/intrinsic.rs

```diff
@@ -138,6 +141,18 @@ pub fn trans_intrinsic_call<'a, 'tcx>(bx: &Builder<'a, 'tcx>,
             let llfn = cx.get_intrinsic(&("llvm.debugtrap"));
             bx.call(llfn, &[], None)
         }
+        "va_arg" => {
+            match ret_ty.sty {
+                // `va_arg` should never be used with the return type f32. If the
+                // return type is f32, use `va_arg` with the return type of f64
+                // followed by fptrunc to truncate the result to an f32.
+                ty::TyFloat(ast::FloatTy::F32) => {
```

Comment:
@eddyb 

> At best, we can generate special code for it, to make it work, in the intrinsic.

Is this kind of what you're looking for?

A quick test shows that it emits.

```
%6 = va_arg i64* %5, double, !dbg !31
%7 = fptrunc double %6 to float, !dbg !31
store float %7, float* %tmp_ret, align 4, !dbg !31
```

Reply:
Looks good to me. cc @rkruppe

Reply:
Oh, I just realized, you want to check `fn_ty.ret.layout.abi` for `Scalar` with an `F32` inside.
That's because `#[repr(transparent)]` might be involved.